process_video.py

```diff
@@ -0,0 +1,243 @@
+    total_process_time = 0 
+    current_frame_index = 0
+    mask_output_root = cfg['output_dir']
+    saver = ResultSaver(mask_output_root,
+                        '',
+                        dataset='',
+                        object_manager=processor.object_manager,
+                        use_long_id=use_long_id,
+                        palette=palette)
+
+    with torch.inference_mode():
+        with torch.amp.autocast(device, enabled=use_amp):
+            pbar = tqdm(total=total_frame_count)
+            pbar.set_description(f'Processing video {video}')
+            while(cap.isOpened()):   
+                # load frame-by-frame
+                _, frame = cap.read()
+                if frame is None or current_frame_index > total_frame_count:
+                    break
+
+                # timing start
+                if 'cuda' in device:
+                    torch.cuda.synchronize(device)
+                    start = torch.cuda.Event(enable_timing=True) if device == 'cuda' else None
+                    end = torch.cuda.Event(enable_timing=True) if device == 'cuda' else None
+                    start.record()
+                else:
+                    a = perf_counter()
+                
+                frame_name = f'{current_frame_index:07d}' + '.png'
+
+                # check if we have a mask to load 
+                mask = None
+                mask_path = path.join(mask_dir, frame_name)
+                if path.exists(mask_path):
+                    mask = Image.open(mask_path)                
+
+                # convert numpy array to pytorch tensor format
+                frame_torch = image_to_torch(frame, device=device)
+                if mask is not None:
+                    # initialize with the mask
+                    mask_np = np.array(mask)
+                    mask_torch = index_numpy_to_one_hot_torch(mask_np, num_objects+1).to(device)
+                    # the background mask is fed into the model
+                    prob = processor.step(frame_torch, mask_torch[1:], idx_mask=False)
+                else:
+                    # propagate only
+                    prob = processor.step(frame_torch)
+
+                # timing end
+                if 'cuda' in device:
+                    end.record()
+                    torch.cuda.synchronize(device)
+                    total_process_time += (start.elapsed_time(end) / 1000)                        
+                else:
+                    b = perf_counter()
+                    total_process_time += (b - a)
+
+                saver.process(prob,
+                                frame_name,
+                                resize_needed=False,
+                                shape=None,
+                                last_frame=(current_frame_index == total_frame_count - 1),
+                                path_to_image=None)
+                
+                check_to_clear_non_permanent_cuda_memory(processor=processor, device=device)
```

Comment:
I don't see an actual OOM. Which is very nice that that processing do not crash.
But I see issues when max GPU memory is reached.
For context I'm working on Windows 11, using a GeForce GTX 1060 6GB. Testing on a ~1 min (1628 frames) Full HD video.

With out the memory clean up the full video FPS is 3.6.
The ~1200 frames processes smoothly and even, but the I can observ that GPU mem max is reached and processing slows down and is herafter done in bursts very slowly. Seems like some kind of memory swapping happening. But just a guess; I don't know.

With the memory clean up the full video FPS is 8.8.
And all processing runs smoothly and even for all frames. On a GPU mem graph I can observ a drop when the clean up kicks in. But processing time do not really seem to be affected.

The clean up step was inspired by the interactive UI fuctionallity to do the same manually.

Reply:
I see. I don't think PyTorch supports GPU memory swapping so it might just be PyTorch greedily grabbing too much memory. Cleaning up the memory sometimes does improve the speed but might lead to non-intuitive failure (e.g., in some more challenging tracking scenarios). Is it possible to turn this feature into a command line option that is defaulted off?

Reply:
Yes, I will make the memory cleaning optional with a command line arg. 
But will first have time to do this end of next week.

Reply:
I have now added a command line option arg for the cuda memory cleaning. Default disabled.
Let me know what you think.